listenbrainz/webserver/static/js/src/RecentListens.test.tsx

```diff
@@ -1111,3 +1111,120 @@ describe("Pagination", () => {
     });
   });
 });
+
+describe("handleClickNewerRecommendations", () => {
+  it("don't do anything if already on newest/first page i.e. offset === 0", async () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    const recs = recentListensProps.listens.slice(0, 25);
+    wrapper.setState({ offset: 0, listens: recs });
+
+    await instance.handleClickNewerRecommendations();
+
+    expect(wrapper.state("offset")).toEqual(0);
+    expect(wrapper.state("listens")).toEqual(recs);
+  });
+
+  it("go to the previous page and update listens if offset !== 0", async () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    wrapper.setState({
+      offset: 50,
+      listens: recentListensProps.listens.slice(50, 53),
+    });
+
+    await instance.handleClickNewerRecommendations();
+
+    expect(wrapper.state("offset")).toEqual(25);
+    expect(wrapper.state("listens").length).toEqual(25);
+  });
+});
+
+describe("handleClickNewestRecommendations", () => {
+  it("don't do anything if already on newest/first page i.e. offset === 0", async () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    const recs = recentListensProps.listens.slice(0, 25);
+    wrapper.setState({ offset: 0, listens: recs });
+
+    await instance.handleClickNewestRecommendations();
+
+    expect(wrapper.state("offset")).toEqual(0);
+    expect(wrapper.state("listens")).toEqual(recs);
+  });
+
+  it("move to the newest/first page if offset !== 0", async () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    wrapper.setState({
+      offset: 50,
+      listens: recentListensProps.listens.slice(50, 53),
+    });
+
+    await instance.handleClickNewestRecommendations();
+
+    expect(wrapper.state("offset")).toEqual(0);
+    expect(wrapper.state("listens").length).toEqual(25);
+  });
+});
+
+describe("handleClickOldestRecommendations", () => {
+  it("don't do anything if already on the last page", async () => {
+    const wrapper = shallow<RecentListens>(<RecentListens {...props} />);
+    const instance = wrapper.instance();
+
+    const recs = recentListensProps.listens.slice(50, 53);
+    wrapper.setState({ offset: 50, listens: recs, recommendationLength: 52 });
+
+    await instance.handleClickOldestRecommendations();
+
+    expect(wrapper.state("offset")).toEqual(50);
+    expect(wrapper.state("listens")).toEqual(recs);
+  });
+  /*
+  it("move to the last page", async () => {
```

Comment:
The commented tests fail for some reason which I don't understand.
Error: 

`Summary of all failing tests
 FAIL  static/js/src/RecentListens.test.tsx
  ● Test suite failed to run

    RangeError: Maximum call stack size exceeded

      at Math.<anonymous> (node_modules/jest-mock/build/index.js:379:51)
      at randomIntInRange (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:43:33)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:75:22)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
      at doQuickSort (node_modules/source-map-support/node_modules/source-map/lib/quick-sort.js:100:5)
`

Reply:
This sounds like runaway recursion....

Reply:
Thanks :)